[Feature] Automatic tap changer documentation (#606)
PowerGridModel/power-grid-model: docs/user_manual/calculations.md

| ------------------------------------------------------------ | -------- | -------- | ----------------------------------------------------------------------------------------------------------- |
| No automatic tap changing (regular power flow)               | &#10004; | &#10004; | {py:class}`TapChangingStrategy.disabled <power_grid_model.enum.TapChangingStrategy.disabled>`               |
| Optimize tap position for any value in the voltage band      |          | &#10004; | {py:class}`TapChangingStrategy.any_valid_tap <power_grid_model.enum.TapChangingStrategy.any_valid_tap>`     |
| Optimize tap position for lower voltage in the voltage band  |          |          | {py:class}`TapChangingStrategy.min_voltage_tap <power_grid_model.enum.TapChangingStrategy.min_voltage_tap>` |
| Optimize tap position for higher voltage in the voltage band |          |          | {py:class}`TapChangingStrategy.max_voltage_tap <power_grid_model.enum.TapChangingStrategy.max_voltage_tap>` |

The following control logic is used.

> - Regulated transformers are ranked according to the amount of transformers between itself and the nearest source {hoverxreftooltip}`user_manual/components:source`.

Reviewer: Maybe good to mention that they are ranked based on voltage level.

Author: what do you mean? they are regulated based on the weighted shortest path to the source, right? Or do you mean that only paths from the higher voltage to the same or lower rated voltage are taken into account?

Reviewer: I meant that it is not necessarily the amount of transformers: if you have a transformer that has similar rated voltage on both sides, it does not count in the ranking process

Author: It does. Taken from the mathematical workout of the ranking algorithm:

>    2. For each non-transformer branch (link, line) which are connected at both sides 
>       1. Create two edges for both direction, with the weight zero. 
>    3. **For each transformer which are connected at both sides** 
>       1. **If there is no automatic tap changer** 
>          1. **Create two edges for both direction, with the weight one.** 
>       2. **If there is automatic tap changer** 
>          1. **Create one edge from the node at the tap side to the node without tap, with the weight one.** 

Lines are not taken into account, but all transformers are. The transformers with similar `u_rated` on both sides are just taken into account bidirectionally, rather than unidirectionally. That means that in the following grid

Author: cfr. discussion: indeed correct as is

    - Loop over all ranks:
      - Loop over all transformers within this rank; transformers with the same rank are independently regulated:
        - If the $U_{\text{control}} < U_{\text{set}} - \frac{U_{\text{band}}}{2}$:
          - The ratio between the voltage on the tap side and the voltage on the control side is too high.
          - To decrease this ratio, the tap ratio must be decreased.
          - Therefore, the tap position of the regulated transformer is decreased if it satisfies the bounds set by `tap_min` and `tap_max`.
        - If, however, the $U_{\text{control}} > U_{\text{set}} + \frac{U_{\text{band}}}{2}$:
          - The ratio between the voltage on the tap side and the voltage on the control side is too low.
          - To increase this ratio, the tap ratio must be increase.
          - Therefore, the tap position of the regulated transformer is increased if it satisfies the bounds set by `tap_min` and `tap_max`.
        - If the tap position of this transformer did not change, the transformer is considered regulated.

>       - If not all transformers within this rank are regulated, step out of the loop and go to the next iteration step.

Reviewer: Would be good to mention when this would happen

  - Exploit the neighbourhood of all transformers (see {hoverxreftooltip}`user_manual/calculations:Initialization and exploitation of regulated transformers`)
    - Re-run the iteration in the above if any of the tap positions changed by the exploitation.

The exploitation of the neighbourhood ensures that the actual optimum is not accidentally missed due to feedback mechanisms in the grid.

> ```{note}

Reviewer: This is a general note for all PF calculations right? Might be confusing that is is within a section about "Power flow with automatic tap changing"

For iterative [power flow calculation methods](#power-flow-algorithms), the initial state may not converge.
If the iterative process failed to converge, the optimization procedure is executed twice.
First, the {py:class}`linear <power_grid_model.enum.CalculationMethod.linear>` calculation method is used to find an approximate solution.
The optimization procedure is then run a second time to find the actual optimum with the user-specified calculation method.
```

```{note}
The control logic assumes that the change in voltage level caused by changing a transformer is larger than the change caused by adjacent transformers.
This is assumption is reflected in the requirements mentioned in {hoverxreftooltip}`user_manual/components:Transformer Tap Regulator`.
```

```{note}
The control logic assumes that the (compensated) control voltage decreases when the tap position increases.
If the line drop compensation impedance is high, and the control side has generator-like behavior, then this assumption does not hold, and the calculation may diverge.

> Hence, this is assumption is reflected in the requirements mentioned in {hoverxreftooltip}`user_manual/components:Line drop compensation`.

Reviewer: ```suggestion
Hence, this assumption is reflected in the requirements mentioned in {hoverxreftooltip}`user_manual/components:Line drop compensation`.
```

        - If the tap position of this transformer did not change, the transformer is considered regulated.
      - If not all transformers within this rank are regulated:
        - A better combination of tap positions may have been found.
        - Step out of the loop and go to the next iteration step.
  - Exploit the neighbourhood of all transformers (see {hoverxreftooltip}`user_manual/calculations:Initialization and exploitation of regulated transformers`)
    - Re-run the iteration in the above if any of the tap positions changed by the exploitation.

The exploitation of the neighbourhood ensures that the actual optimum is not accidentally missed due to feedback mechanisms in the grid.

```{note}
For iterative [power flow calculation methods](#power-flow-algorithms), the initial state may not converge.
If the iterative process failed to converge, the optimization procedure is executed twice.
First, the {py:class}`linear <power_grid_model.enum.CalculationMethod.linear>` calculation method is used to find an approximate solution.
The optimization procedure is then run a second time to find the actual optimum with the user-specified calculation method.
```

```{note}

> The control logic assumes that the change in voltage level caused by changing a transformer is larger than the change caused by adjacent transformers.

Reviewer 2: ```suggestion
The control logic assumes that the change in voltage level caused by regulating a transformer is larger than the change caused by adjacent transformers.
```
Or maybe something else is meant

This is assumption is reflected in the requirements mentioned in {hoverxreftooltip}`user_manual/components:Transformer Tap Regulator`.
```

```{note}
The control logic assumes that the (compensated) control voltage decreases when the tap position increases.
If the line drop compensation impedance is high, and the control side has generator-like behavior, then this assumption does not hold, and the calculation may diverge.
Hence, this assumption is reflected in the requirements mentioned in {hoverxreftooltip}`user_manual/components:Line drop compensation`.
```

##### Initialization and exploitation of regulated transformers

> Regulated transformers are initialized and exploited depending on the optimization strategy.

Reviewer 2: In the context of all written text about `exploitation`, I think more information can be added here or below: 

Above the table,
Internally, the control algorithm sets initial tap position as following so we end up with a regulated tap_pos corresponding to the following conditions:

In the description section
`Find best possible tap position that gives highest control side voltage within the u_band.`
`Find best possible tap position that gives lowest control side voltage within the u_band.`

Tap side voltage seems less relevant i suppose.

| ------------------------------------------------------------ | -------- | -------- | ----------------------------------------------------------------------------------------------------------- |
| No automatic tap changing (regular power flow)               | &#10004; | &#10004; | {py:class}`TapChangingStrategy.disabled <power_grid_model.enum.TapChangingStrategy.disabled>`               |
| Optimize tap position for any value in the voltage band      |          | &#10004; | {py:class}`TapChangingStrategy.any_valid_tap <power_grid_model.enum.TapChangingStrategy.any_valid_tap>`     |
| Optimize tap position for lower voltage in the voltage band  |          |          | {py:class}`TapChangingStrategy.min_voltage_tap <power_grid_model.enum.TapChangingStrategy.min_voltage_tap>` |
| Optimize tap position for higher voltage in the voltage band |          |          | {py:class}`TapChangingStrategy.max_voltage_tap <power_grid_model.enum.TapChangingStrategy.max_voltage_tap>` |

##### Control logic for power flow with automatic tap changing

The following control logic is used.

> - Regulated transformers are ranked according to the amount of transformers between itself and the nearest source {hoverxreftooltip}`user_manual/components:source`.

Author: the user doesn't care what is low and what is high, though. i agree the original was incomplete but i feel like your proposal also doesn't fully capture the info that is relevant to the user. i'll think about it for a second but if you have another proposal, please LMK

> - Regulated transformers are ranked according to the amount of transformers between itself and the nearest source {hoverxreftooltip}`user_manual/components:source`.

Author: maybe (including the next line):

```md
- Regulated transformers are prioritized according to the amount of transformers to a {hoverxreftooltip}`user_manual/components:source` in a transformer-source topological graph.
  - Transformers that are directly connected to a {hoverxreftooltip}`user_manual/components:source` are prioritized for regulation.
```

> - Regulated transformers are ranked according to the amount of transformers between itself and the nearest source {hoverxreftooltip}`user_manual/components:source`.

Reviewer 3: In your own words: 
```
  - Transformers that are directly connected to a {hoverxreftooltip}`user_manual/components:source` are prioritized for regulation. 
```
... this captures even less ;)

> - Regulated transformers are ranked according to the amount of transformers between itself and the nearest source {hoverxreftooltip}`user_manual/components:source`.

Reviewer 3: second line:
```
 - Transformers are regulated in order according to their ranks.
```

>   - Transformers that are closer to a {hoverxreftooltip}`user_manual/components:source` are regulated first.

Reviewer 3: ```suggestion
  - Transformers of lower ranks are prioritized for regulation before those of higher ranks.
```

- Initialize all transformers to their starting tap position (see {hoverxreftooltip}`user_manual/calculations:Initialization and exploitation of regulated transformers`)
- Find the optimal state using the following procedure

>   - While not all transformers are regulated, iterate as follows:

Reviewer 3: This sounds like there is a state saying if a transformer is regulated already.
```suggestion
  - While not all transformers cannot be further regulated, iterate as follows:
```
or
```suggestion
  - While some transformers can still be further regulated, iterate as follows:
```

    - Run a power flow calculation with the current tap positions with the specified [calculation method](#power-flow-algorithms).

>     - Start with the lowest rank (with the transformers closest to a {hoverxreftooltip}`user_manual/components:source`, because the source provides a relatively stable voltage level).

Reviewer 3: I thought the users *don't care* about low or high? ;)

Author: indeed, will remove

>     - Loop over all ranks:

Author: the two are related. will update
